index.js

```diff
@@ -0,0 +1,214 @@
+import * as ethers from 'ethers'
+import { json, status } from 'http-responders'
+import assert from 'node:assert'
+
+const maxScore = BigInt(1e15)
+// https://github.com/filecoin-station/spark-impact-evaluator/blob/fd64313a96957fcb3d5fda0d334245601676bb73/test/Spark.t.sol#L11C39-L11C65
+const roundReward = 456621004566210048n
+
+const handler = async (req, res, redis, signerAddresses) => {
+  if (req.method === 'POST' && req.url === '/scores') {
+    await handleIncreaseScores(req, res, redis, signerAddresses)
+  } else if (req.method === 'POST' && req.url === '/paid') {
+    await handlePaidScheduledRewards(req, res, redis, signerAddresses)
+  } else if (req.method === 'GET' && req.url === '/scheduled-rewards') {
+    await handleGetAllScheduledRewards(res, redis)
+  } else if (req.method === 'GET' && req.url.startsWith('/scheduled-rewards/')) {
+    await handleGetSingleScheduledRewards(req, res, redis)
+  } else if (req.method === 'GET' && req.url === '/log') {
+    await handleGetLog(res, redis)
+  } else {
+    status(res, 404)
+  }
+}
+
+const validateSignature = (signature, participants, signerAddresses) => {
+  httpAssert(
+    typeof signature === 'object' && signature !== null,
+    400,
+    '.signature should be an object'
+  )
+  httpAssert.deepEqual(
+    Object.keys(signature).sort(),
+    ['r', 's', 'v'],
+    400,
+    '.signature should have keys .r, .s and .v'
+  )
+
+  const digest = ethers.solidityPackedKeccak256(
+    ['address[]', 'int256[]'],
+    [Object.keys(participants), Object.values(participants)]
+  )
+  const reqSigner = ethers.verifyMessage(
+    digest,
+    ethers.Signature.from(signature)
+  )
+  httpAssert(signerAddresses.includes(reqSigner), 403, 'Invalid signature')
+}
+
+async function handleIncreaseScores (req, res, redis, signerAddresses) {
+  const body = JSON.parse(await getRawBody(req, { limit: '1mb' }))
+
+  httpAssert(
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.scores === 'object' && body.scores !== null,
+    400,
+    '.scores should be an object'
+  )
+  httpAssert(
+    Object.keys(body.scores).every(ethers.isAddress),
+    400,
+    'All .scores keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.scores).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .scores values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.scores, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, score] of Object.entries(body.scores)) {
+    const scheduledRewards = (BigInt(score) * roundReward) / maxScore
+    tx.hincrby('rewards', address, scheduledRewards)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        score,
+        scheduledRewards: String(scheduledRewards)
```

Comment:
When I read the docs, I assumed `scheduledRewards` contained the new amount after the round rewards were applied. It took me a while to notice that the log tells us only the delta.

Let's find a better name for this field, please.

For example: `rewardsChange`, `rewardsDelta`

Since this service is all about scheduled rewards, maybe we don't need to mention rewards and use `balance` instead, e.g. `balanceChange`.

Reply:
Good idea, I got confused myself about this too. I changed it to `scheduledRewardsDelta` - it's clear and explicit, and in this case it doesn't really matter how long the name is